Upgrade `react-router-dom` to version 6

mashikag opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->
This PR affects the following. I did not deal with most of that codebase before, hence I am asking for your help in verifying everything works as expected.
- `stepper`
   - @alshakero can you please help me to make sure there are no regressions? 🙏 
- `onboarding`
  - @alshakero can you please help me to make sure there are no regressions? 🙏 
- `help-center`
  - @alshakero, @arcangelini, @escapemanuele, @candy02058912  can you please help me to make sure there are no regressions? 🙏 

## Proposed Changes

* Upgrade `react-router-dom` library to the latest version

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test for no regressions

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

mashikag committed:
Upgrage react-router-dom to version 6

mashikag committed:
Revert unnecessary comment amendment

mashikag committed:
Fix stepper navigation

mashikag committed:
temp

mashikag committed:
Resolve help center routes

mashikag committed:
Use useNavigate instead of Navigate component

mashikag committed:
Fix onboarding tests

mashikag committed:
Fix help-center tests

mashikag committed:
Fix stepper tests

mashikag committed:
Dedupe and update yarn.lock

alshakero commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -177,42 +206,30 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
-			</Switch>
+			<Routes>
+				{ flowSteps.map( ( step ) => (
+					<Route
+						key={ step.slug }
+						path={ `/${ flow.variantSlug ?? flow.name }/${ step.slug }` }
+						element={
```
Maybe move this inside `StepRoute`?

alshakero commented on packages/help-center/src/components/help-center-sibyl-articles.tsx:
```diff
@@ -114,9 +114,7 @@ export function SibylArticles( {
 				...article,
-				url: hasPostId
-					? getPostUrl( article as Article, message, articleCanNavigateBack )
-					: article.link,
+				url: getPostUrl( article as Article, message, articleCanNavigateBack ) ?? article.link,
```
Is this change relevant?

mashikag replied:
Yes, it is relevant. The `url` can no longer be `undefined` on the `ConfigureableLink` component. The `ConfigureableLink` component extends its props with `LinkProps` from `react-router-dom`. 

Also, if you were to study the return value of the conditional assignment, it will not resolve to `undefined`. The `getPostUrl` function, checks whether the article prop has an ID.  It will return `undefined` if the article does not have an ID. We already ensure that the article passed in has an ID with the `hasPostId` const. However, typescript is not aware of this, and hence it believes the return type can still be `undefined` when an article has an ID.

![Screen Shot 2023-04-05 at 15 34 57](https://user-images.githubusercontent.com/2019970/230096906-11858c78-d158-4dd5-b799-6c4644912abe.png)

alshakero reviewed (changes requested):
Stepper looks good. The Help Center's API is broken, though. 

1. The testing steps [here](https://github.com/Automattic/wp-calypso/pull/68639) will fail. 
2. I don't have testing steps for this case, but [this](https://github.com/Automattic/wp-calypso/blob/update/react-router-dom-upgrade/client/state/jitm/actions.js#L100) will fail as well. 

We have more potential uses for this `MemoryRouter` feature that we wanted to use in the future, namely persisting the Help Center router state across sessions, but that's just a potential use that we can drop. The showstopper is the missing ability to open the Help Center at a given route. We can maybe create a new state in the store called `initialRoute` or something and redirect to that if it's set.

mashikag committed:
Move StepRoute component into its own file

mashikag commented:
> Stepper looks good. The Help Center's API is broken, though.
> 
> 1. The testing steps [here](https://github.com/Automattic/wp-calypso/pull/68639) will fail.

As far as I understand the testing instructions, my PR does not break them. The chat window still opens without storing the router state because the `data.status` is `'assigned', which is causing the route change to `/inline-chat?session=continued`.

mashikag committed:
Introduce initialRoute state into help-center data-store

mashikag committed:
Update yarn.lock and bring back change that was reverted by mistake in my last commit

mashikag committed:
Merge latest trunk

mashikag committed:
Remove empty launch files

mashikag committed:
Remove empty scroll-to-top/test/index.tsx files - this was left after conflicts resolution

mashikag commented:
Hey @alshakero! Thanks again for the video call we had earlier on to discuss my changes, and your review. It was very helpful! 🙇‍♂️

> Stepper looks good. The Help Center's API is broken, though.
> 
> 1. The testing steps [here](https://github.com/Automattic/wp-calypso/pull/68639) will fail.
> 2. I don't have testing steps for this case, but [this](https://github.com/Automattic/wp-calypso/blob/update/react-router-dom-upgrade/client/state/jitm/actions.js#L100) will fail as well.

I have now addressed the two points. I introduced the `initialRoute` state variable in the `HELP_CENTER` data store, just as we agreed on the Zoom call we had.

Furthermore, I would only seek for an approval from @candy02058912 that what I did to [`client/state/jitm/actions.js`](https://github.com/Automattic/wp-calypso/pull/75283/files#diff-f152ca70bb0ebe14f9708317690ad5f6da410810f5351ff5753bc94dec4a7354) works as intended. 🙏

alshakero reviewed (approved):
Tested Stepper and the Help Center and all works as intended!
